hoomd/state.py

```diff
@@ -55,52 +75,54 @@ def __init__(self, simulation, snapshot):
         else:
             self._cpp_sys_def = _hoomd.SystemDefinition(
                 snapshot._cpp_obj, simulation.device._cpp_exec_conf)
+
+        # Necessary for local snapshot API. This is used to ensure two local
+        # snapshots are not contexted at once.
+        self._in_context_manager = False
+
+        # provides a cache of C++ group objects of the form {type(filter):
+        # {filter: C++ group}}. The first layer is to prevent user created
+        # filters with poorly implemented __hash__ and __eq__ from causing cache
+        # errors.
         self._groups = defaultdict(dict)
 
     @property
     def snapshot(self):
-        cpp_snapshot = self._cpp_sys_def.takeSnapshot_double()
-        return Snapshot._from_cpp_snapshot(cpp_snapshot,
-                                           self._simulation.device.communicator)
+        """hoomd.Snapshot: All data of a simulation's current microstate.
```

Comment:
We need to document the `hoomd.Snapshot` object in detail, though that should be the subject of another PR. If use of `Snapshot` is as discouraged as these docs suggest, should we move the `Snapshot` class into `data` in our namespace reorg?

Reply:
I didn't mean to discourage its use, I more was trying to prevent users from doing something like
````python
snapshot = sim.state.snapshot
if snapshot.exists:
    snapshot.particles.velocities[:] = [0, 0, 0]
# or
if snapshot.exists:
    snapshot.particles.position[1] = [1, 2, 3]
sim.state.snapshot = snapshot
````
Overall, I think that the plain snapshot is easiest to manipulate especially given an MPI simulation. I just wanted to inform users that if only manipulating a few pieces of data or doing some blanket change like above, it is faster to use the local snapshot API.

I am not sure what the right balance here between warning users about using a sub-optimal method (for some use cases) versus the ease in use of the snapshot.

Reply:
In reference docs, I would prefer to state facts and leave suggested usages for tutorials. E.g., we could state that `snapshot` is an O(N_particles + N_bonds + ...) operation that provides a copy of the data while the local snapshot methods are nominally O(1) and provides a reference.